File: ts/src/providers/openai.ts
import OpenAI from "openai";
import type {
  ChatCompletion,
  ChatCompletionChunk,
  ChatCompletionMessageParam,
} from "openai/resources/chat/completions";
import { AnyLLM } from "../any_llm.js";
import {
  AuthenticationError,
  ContextLengthExceededError,
  InvalidRequestError,
  MissingApiKeyError,
  ModelNotFoundError,
  ProviderError,
  RateLimitError,
} from "../errors.js";
import type { CompletionOptions, CompletionParams } from "../types.js";

const OPENAI_API_KEY_ENV = "OPENAI_API_KEY";
const PROVIDER_NAME = "openai";

/**
 * OpenAI provider implementation for AnyLLM.
 */
export class OpenAIProvider extends AnyLLM {
  readonly providerName = PROVIDER_NAME;

  private client!: OpenAI;

  /**
   * Initialize the OpenAI client.
   * Falls back to OPENAI_API_KEY environment variable if apiKey is not provided.
   *
   * @param apiKey - Optional API key
   * @param apiBase - Optional API base URL
   * @param options - Additional options (e.g., defaultHeaders)
   */
  initClient(
    apiKey?: string,
    apiBase?: string,
    options?: Record<string, unknown>,
  ): void {
    const resolvedApiKey = apiKey ?? process.env[OPENAI_API_KEY_ENV];

    if (!resolvedApiKey) {
      throw new MissingApiKeyError(PROVIDER_NAME, OPENAI_API_KEY_ENV);
    }

    this.client = new OpenAI({
      apiKey: resolvedApiKey,
      baseURL: apiBase,
      defaultHeaders: options?.defaultHeaders as
        | Record<string, string>
        | undefined,
    });
  }

  /**
   * Perform a chat completion request.
   */
  async completion(
    model: string,
    messages: ChatCompletionMessageParam[],
    options?: CompletionOptions,
  ): Promise<ChatCompletion> {
    const params = this.buildCompletionParams(model, messages, options, false);
    const requestParams = OpenAIProvider.toOpenAIParams(params);

    try {
      const response = await this.client.chat.completions.create({
        ...requestParams,
        stream: false,
      });
      return response;
    } catch (error) {
      throw this.mapError(error);
    }
  }

  /**
   * Perform a streaming chat completion request.
   */
  async *completionStream(
    model: string,
    messages: ChatCompletionMessageParam[],
    options?: CompletionOptions,
  ): AsyncIterable<ChatCompletionChunk> {
    const params = this.buildCompletionParams(model, messages, options, true);
    const requestParams = OpenAIProvider.toOpenAIParams(params);

    try {
      const stream = await this.client.chat.completions.create({
        ...requestParams,
        stream: true,
      });

      for await (const chunk of stream) {
        yield chunk;
      }
    } catch (error) {
      throw this.mapError(error);
    }
  }

  /**
   * Convert CompletionParams to OpenAI-specific request parameters.
   * For OpenAI, the params are already in the correct format.
   */
  private static toOpenAIParams(
    params: CompletionParams,
  ): OpenAI.Chat.ChatCompletionCreateParams {
    // OpenAI uses the standard format, so we can pass through most params
    // Build the params object with only defined values
    const result: OpenAI.Chat.ChatCompletionCreateParams = {
      model: params.model,
      messages: params.messages,
    };

    if (params.temperature !== undefined)
      result.temperature = params.temperature;
    if (params.top_p !== undefined) result.top_p = params.top_p;
    if (params.max_tokens !== undefined) result.max_tokens = params.max_tokens;
    if (params.max_completion_tokens !== undefined)
      result.max_completion_tokens = params.max_completion_tokens;
    if (params.n !== undefined) result.n = params.n;
    if (params.stop !== undefined) result.stop = params.stop;
    if (params.presence_penalty !== undefined)
      result.presence_penalty = params.presence_penalty;
    if (params.frequency_penalty !== undefined)
      result.frequency_penalty = params.frequency_penalty;
    if (params.logit_bias !== undefined) result.logit_bias = params.logit_bias;
    if (params.user !== undefined) result.user = params.user;
    if (params.tools !== undefined) result.tools = params.tools;
    if (params.tool_choice !== undefined)
      result.tool_choice = params.tool_choice;
    if (params.parallel_tool_calls !== undefined)
      result.parallel_tool_calls = params.parallel_tool_calls;
    if (params.response_format !== undefined) {
      result.response_format =
        params.response_format as OpenAI.Chat.ChatCompletionCreateParams["response_format"];
    }
    if (params.logprobs !== undefined) result.logprobs = params.logprobs;
    if (params.top_logprobs !== undefined)
      result.top_logprobs = params.top_logprobs;
    if (params.stream_options !== undefined)
      result.stream_options = params.stream_options;

    return result;
  }

  /**
   * Map OpenAI errors to standardized AnyLLM errors.
   */
  private mapError(error: unknown): Error {
    if (!(error instanceof Error)) {
      return new ProviderError(String(error), undefined, PROVIDER_NAME);
    }

    // Handle OpenAI SDK errors
    if (error instanceof OpenAI.APIError) {
      const status = error.status;
      const message = error.message;

      // Rate limit errors (429)
      if (status === 429) {
        return new RateLimitError(message, error, PROVIDER_NAME);
      }

      // Authentication errors (401)
      if (status === 401) {
        return new AuthenticationError(message, error, PROVIDER_NAME);
      }

      // Invalid request errors (400)
      if (status === 400) {
        // Check for context length errors
        if (
          message.toLowerCase().includes("context length") ||
          message.toLowerCase().includes("maximum context") ||
          message.toLowerCase().includes("token limit")
        ) {
          return new ContextLengthExceededError(message, error, PROVIDER_NAME);
        }
        return new InvalidRequestError(message, error, PROVIDER_NAME);
      }

      // Not found errors (404)
      if (status === 404) {
        // Check for model not found
        if (message.toLowerCase().includes("model")) {
          return new ModelNotFoundError(message, error, PROVIDER_NAME);
        }
        return new InvalidRequestError(message, error, PROVIDER_NAME);
      }

      // Server errors (5xx)
      if (status && status >= 500) {
        return new ProviderError(message, error, PROVIDER_NAME);
      }

      // Other API errors
      return new ProviderError(message, error, PROVIDER_NAME);
    }

    // Generic error handling
    return new ProviderError(error.message, error, PROVIDER_NAME);
  }
}

File: rs/src/providers/openai.rs
//! `OpenAI` provider implementation.
//!
//! This module provides the `OpenAI` provider for the any-llm library.

use std::collections::HashMap;
use std::env;
use std::pin::Pin;

use async_stream::try_stream;
use futures::Stream;
use reqwest::Client;
use reqwest::header::{AUTHORIZATION, CONTENT_TYPE, HeaderMap, HeaderName, HeaderValue};
use serde::{Deserialize, Serialize};

use crate::errors::{AnyLLMError, Result};
use crate::types::{ChatCompletion, ChatCompletionChunk, CompletionOptions, Message};

/// Default `OpenAI` API base URL.
const DEFAULT_API_BASE: &str = "https://api.openai.com/v1";

/// Environment variable name for `OpenAI` API key.
const ENV_VAR_NAME: &str = "OPENAI_API_KEY";

/// Provider name.
const PROVIDER_NAME: &str = "OpenAI";

/// `OpenAI` completion request body.
#[derive(Debug, Serialize)]
struct CompletionRequest {
    model: String,
    messages: Vec<Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_p: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    max_completion_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    n: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stop: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    presence_penalty: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    frequency_penalty: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    logit_bias: Option<HashMap<String, f64>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    logprobs: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_logprobs: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    seed: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    user: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tools: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tool_choice: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    parallel_tool_calls: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    response_format: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream_options: Option<serde_json::Value>,
}

/// `OpenAI` API error response.
#[derive(Debug, Deserialize)]
struct ApiErrorResponse {
    error: ApiError,
}

/// `OpenAI` API error details.
#[derive(Debug, Deserialize)]
struct ApiError {
    message: String,
    #[serde(rename = "type")]
    error_type: Option<String>,
    code: Option<String>,
}

/// `OpenAI` provider client.
#[derive(Debug)]
pub struct OpenAIProvider {
    client: Client,
    api_base: String,
    headers: HeaderMap,
}

impl OpenAIProvider {
    /// Create a new `OpenAI` provider.
    ///
    /// # Arguments
    ///
    /// * `api_key` - API key. If None, reads from `OPENAI_API_KEY` environment variable.
    /// * `api_base` - API base URL. If None, uses the default `OpenAI` URL.
    /// * `extra_headers` - Additional headers to include in requests.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is provided and the environment variable is not set.
    pub fn new(
        api_key: Option<String>,
        api_base: Option<String>,
        extra_headers: Option<HashMap<String, String>>,
    ) -> Result<Self> {
        let api_key = api_key
            .or_else(|| env::var(ENV_VAR_NAME).ok())
            .ok_or_else(|| AnyLLMError::missing_api_key(PROVIDER_NAME, ENV_VAR_NAME))?;

        let api_base = api_base.unwrap_or_else(|| DEFAULT_API_BASE.to_string());

        let mut headers = HeaderMap::new();
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        headers.insert(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {api_key}"))
                .map_err(|e| AnyLLMError::invalid_request(PROVIDER_NAME, e.to_string()))?,
        );

        if let Some(extra) = extra_headers {
            for (key, value) in extra {
                let header_name = HeaderName::try_from(&key)
                    .map_err(|e| AnyLLMError::invalid_request(PROVIDER_NAME, e.to_string()))?;
                let header_value = HeaderValue::from_str(&value)
                    .map_err(|e| AnyLLMError::invalid_request(PROVIDER_NAME, e.to_string()))?;
                headers.insert(header_name, header_value);
            }
        }

        let client = Client::new();

        Ok(Self {
            client,
            api_base,
            headers,
        })
    }

    /// Convert completion options to request parameters.
    fn convert_completion_params(
        model: &str,
        messages: Vec<Message>,
        options: Option<CompletionOptions>,
        stream: bool,
    ) -> Result<CompletionRequest> {
        let options = options.unwrap_or_default();

        let tools = options.tools.map(serde_json::to_value).transpose()?;
        let tool_choice = options.tool_choice.map(serde_json::to_value).transpose()?;
        let response_format = options
            .response_format
            .map(serde_json::to_value)
            .transpose()?;
        let stream_options = if stream {
            options
                .stream_options
                .map(serde_json::to_value)
                .transpose()?
        } else {
            None
        };

        Ok(CompletionRequest {
            model: model.to_string(),
            messages,
            temperature: options.temperature,
            top_p: options.top_p,
            max_tokens: options.max_tokens,
            max_completion_tokens: options.max_completion_tokens,
            n: options.n,
            stop: options.stop,
            presence_penalty: options.presence_penalty,
            frequency_penalty: options.frequency_penalty,
            logit_bias: options.logit_bias,
            logprobs: options.logprobs,
            top_logprobs: options.top_logprobs,
            seed: options.seed,
            user: options.user,
            tools,
            tool_choice,
            parallel_tool_calls: options.parallel_tool_calls,
            response_format,
            stream: if stream { Some(true) } else { None },
            stream_options,
        })
    }

    /// Map API error to `AnyLLMError`.
    fn map_error(status: u16, error: ApiError) -> AnyLLMError {
        let code = error.code.as_deref().unwrap_or("");
        let error_type = error.error_type.as_deref().unwrap_or("");

        match status {
            401 => AnyLLMError::authentication(PROVIDER_NAME, Some(error.message)),
            429 => AnyLLMError::rate_limit(PROVIDER_NAME, Some(error.message)),
            400 => {
                if code == "context_length_exceeded" || error_type == "context_length_exceeded" {
                    AnyLLMError::context_length_exceeded(PROVIDER_NAME, Some(error.message))
                } else if code == "model_not_found" || error_type == "model_not_found" {
                    AnyLLMError::model_not_found(PROVIDER_NAME, Some(error.message))
                } else {
                    AnyLLMError::invalid_request(PROVIDER_NAME, error.message)
                }
            }
            403 => {
                if code == "content_policy_violation" || error_type == "content_filter" {
                    AnyLLMError::content_filter(PROVIDER_NAME, Some(error.message))
                } else {
                    AnyLLMError::authentication(PROVIDER_NAME, Some(error.message))
                }
            }
            404 => AnyLLMError::model_not_found(PROVIDER_NAME, Some(error.message)),
            _ => AnyLLMError::provider_error(PROVIDER_NAME, Some(error.message)),
        }
    }

    /// Perform a completion request.
    ///
    /// # Arguments
    ///
    /// * `model` - Model ID to use.
    /// * `messages` - Conversation messages.
    /// * `options` - Optional completion parameters.
    ///
    /// # Errors
    ///
    /// Returns an error if the request fails or the response cannot be parsed.
    pub async fn completion(
        &self,
        model: &str,
        messages: Vec<Message>,
        options: Option<CompletionOptions>,
    ) -> Result<ChatCompletion> {
        let request_body = Self::convert_completion_params(model, messages, options, false)?;

        let url = format!("{}/chat/completions", self.api_base);
        let response = self
            .client
            .post(&url)
            .headers(self.headers.clone())
            .json(&request_body)
            .send()
            .await?;

        let status = response.status().as_u16();

        if !response.status().is_success() {
            let error_response: ApiErrorResponse = response.json().await?;
            return Err(Self::map_error(status, error_response.error));
        }

        let completion: ChatCompletion = response.json().await?;
        Ok(completion)
    }

    /// Perform a streaming completion request.
    ///
    /// # Arguments
    ///
    /// * `model` - Model ID to use.
    /// * `messages` - Conversation messages.
    /// * `options` - Optional completion parameters.
    ///
    /// # Errors
    ///
    /// Returns an error if the request fails.
    pub async fn completion_stream(
        &self,
        model: &str,
        messages: Vec<Message>,
        options: Option<CompletionOptions>,
    ) -> Result<Pin<Box<dyn Stream<Item = Result<ChatCompletionChunk>> + Send>>> {
        let request_body = Self::convert_completion_params(model, messages, options, true)?;

        let url = format!("{}/chat/completions", self.api_base);
        let response = self
            .client
            .post(&url)
            .headers(self.headers.clone())
            .json(&request_body)
            .send()
            .await?;

        let status = response.status().as_u16();

        if !response.status().is_success() {
            let error_response: ApiErrorResponse = response.json().await?;
            return Err(Self::map_error(status, error_response.error));
        }

        let stream = try_stream! {
            use futures::StreamExt;

            let mut stream = response.bytes_stream();
            let mut buffer = String::new();

            while let Some(chunk_result) = stream.next().await {
                let chunk = chunk_result?;
                let text = String::from_utf8_lossy(&chunk);
                buffer.push_str(&text);

                while let Some(pos) = buffer.find('\n') {
                    let line = buffer[..pos].to_string();
                    buffer = buffer[pos + 1..].to_string();

                    let line = line.trim();
                    if line.is_empty() {
                        continue;
                    }

                    if let Some(data) = line.strip_prefix("data: ") {
                        if data == "[DONE]" {
                            return;
                        }

                        let chunk: ChatCompletionChunk = serde_json::from_str(data)?;
                        yield chunk;
                    }
                }
            }
        };

        Ok(Box::pin(stream))
    }
}
